Add deprecation subcategories, test deprecations properly including in perldiag and perlderprecation.pod

demerphq opened the pull request:
Currently we use a single warning category for all deprecations. This is dangerous in that it allows a new deprecation to sneak in when the user only wanted to allow a different deprecation, and IMO counterproductive as it means there is no way to know /which/ deprecation someone is silencing in their code and give them good feedback when the feature is no longer supported.  

Having only one category essentially says "if you want to risk using a deprecation then you have to risk not knowing about *any* deprecation", but this does not make sense. Why should I have to take maximal risk when I want to take a known risk?It makes even less sense when you consider we have at least three categories which are long standing and still not scheduled for removal. If a user wants to silence one of these long standing deprecation types they essentially must risk not being informed about any other category at all. That seems unhelpful and unnecessarily risky.

By having all deprecations be subcategorized in regen/warnings.pl we also get some built in support for tracking the version that a deprecation was added, which can then be used to test that our various files provide sufficient dev cycles before they are removed.

It also makes it more obvious that we have unscheduled deprecations that have been around for a very long time, and which will likely always be present (or so it seems).

This PR adds support for deprecation subcategories, it converts all existing deprecations to use subcategories, it fixes the API's of all deprecation related functionality that I could find to require a category, and it ensures that diag.t knows about these functions so it can test the diagnostics produced, it also validates that perldeprecation.pod metions all existing deprecation categories and that they are documented with sufficient version cycles before they are removed. 

I understand that the PSC has "decided" that it is a bad idea to have deprecation subcategories. I believe that this decision is wrong, and should not have been made without wider discusion. This patch demonstrates why it is a good idea to have these subcategories. It facilitates testing and documenting the categories, it allows our users to manage their risk at a more granular level, and it allows us quick access as to which deprecations are "live" at any given point in time. The counter arguments that I have seen expressed, for instance here: https://github.com/Perl/perl5/pull/20855#issuecomment-1445196362 and https://github.com/Perl/perl5/pull/20916https://github.com/Perl/perl5/pull/20855 do not seem very compelling compared to the advantages enumerated here, and demonstrated by this PR.

demerphq committed:
warnings.pm - support deprecated::smartmatch category

Currently we seem to lack a way to have a subcategory under deprecated.
It seems reasonable to me that people might want to disable a specific
subcategory warning while leaving the rest in place. This patch allows
that. Note that both

    no warnings "deprecated";

and

    no warnings "deprecated::smartmatch";

work to disable the warning. Deprecated warnings shouldn't be "all or
nothing", they should be specific and targetted.

demerphq committed:
regcomp_internal.h - make unused macros for deprecation warnings take category as parameter

we should have a deprecation category per type of deprecation

demerphq committed:
warnings.pm - add deprecated::apostrophe_as_package_separator as new deprecation category

This category is about use of apostrophe as a package separator, eg
for things like "Test::More::isn't()".

demerphq committed:
warnings.pm - add deprecated::delimiter_will_be_paired category

Some delimiters are considered deprecated because in the future they
will be used as part of a paired delimiter. This adds a new category
for these cases.

demerphq committed:
utf8.c - add category parameter to unused warn_on_first_deprecated_use function

demerphq committed:
diag.t - detect use of "deprecate_xxx()" style functions

demerphq committed:
embed.fnc - document deprecate_xxx() macros and add them to handy.h

Also do not generate PERL_ARGS style macros for macros.

demerphq committed:
warnings.pm - add deprecated::dot_in_inc warings category

Instead of using a generic warnings category switch to fine grained
control.

demerphq committed:
warnings.pm - support deprecated::unicode_property_name category

This category is only used in the regex engine, we should be able
to disable it specifically, as it seems like we will never actually
remove demove support for the things it warns about.

demerphq committed:
warnings.pm - add deprecated::goto_construct category

This category applies to attempts to goto the internals of a block
construct.

demerphq committed:
warnings.pm - add deprecated::version_downgrade category

This also fixes the version_downgrade to show the correct version that
version downgrades will be removed in.

demerphq committed:
regen/warnings.pl - line up categories, put them in date order

This makes it easier to see the order the categories were added in.

demerphq committed:
perldeprecation.pod - add smartmatch deprecation

This was missed when smartmatch was deprecated.

demerphq committed:
perldeprecation.pod - apostrophe as package separator goes away in 5.42

It was incorrectly documented as going away in 5.40

demerphq committed:
t/porting/deprecation.t - add tests for deprecation documentation and categories

demerphq committed:
perldeprecation.pod - add unscheduled deprecations and add category info

Multiple deprecation types that are not scheduled for removal in a
specific version were not listed. Also now that we have deprecation
subcategories we should specify them in the docs.

leonerd commented on handy.h:
```diff
@@ -2955,18 +2955,39 @@ last-inclusive range.
+=for apidoc Cdm||deprecate_disappears_in|U32 category|"when"|"construct"
+Wrapper around Perl_ck_warner_d() to produce a deprecated warning in the given
+category with an appropriate message that the construct will disappear in a specific
+release. The C<when> and C<construct> arguments must be a C string.
```
It might be useful to document or give an example of the expected format of the `"when"` string, otherwise I can foresee all manner of different styles like `v5.40`, `5.40`, `5.40.0`, `version 5.040`, etc...

demerphq replied:
Good suggestion. Will do. I might even make diag.t test this for us.

demerphq replied:
Added verbiage to document the form, and a test to enforce it.

leonerd commented on regen/warnings.pl:
```diff
@@ -74,7 +74,15 @@ BEGIN
+                                'deprecated::dot_in_inc'               => [ 5.025011, DEFAULT_ON],
+                                'deprecated::version_downgrade'        => [ 5.035009, DEFAULT_ON],
+                                'deprecated::delimiter_will_be_paired' => [ 5.035010, DEFAULT_ON],
+                                'deprecated::apostrophe_as_pack_sep'   => [ 5.037009, DEFAULT_ON],
```
"pack" might be confused with the `pack()` function. Perhaps better to use "pkg" or "package" spelled out in full?

demerphq replied:
I wanted to avoid it getting too long, but given we kind /want/ to discourage people from disabling deprecation warnings (especially this one), I think in hindsight spelling it out makes a lot of sense.

demerphq replied:
Done.

leonerd commented on t/porting/deprecation.t:
```diff
@@ -0,0 +1,145 @@
+my $in_legacy;
+
+while (<$fh>) {
+    if (/=head2 (?|Perl (5\.\d+)(?:\.\d+)?|(Unscheduled))/) { # ignore minor version
```
Tiniest of nitpicks: Should these be `^`-anchored?

demerphq replied:
In practice I dont think it matter that much, but strictly speak yes probably. :-)

demerphq replied:
Done.

leonerd commented on t/porting/deprecation.t:
```diff
@@ -0,0 +1,145 @@
+            $in_legacy = 1;
+        }
+    }
+    if (/=head3 (.*)/) {
```
And also maybe `elsif`, since we don't expect any one line to match more than one of them

demerphq replied:
sure

demerphq replied:
Done.

leonerd reviewed (commented):
Overall this appears to be going the right way; give or take some minor comments below.

Would appreciate a second opinion from @rjbs or @book first though, to check we're aligned with today's re-thinking on PSC.

demerphq committed:
handy.h - fixup for deprecated warnings and docs

demerphq committed:
diag.t - parse and validate "when" parameter from deprecated_xxx() macros

the "when" parameter is expected to be a version string of the form "5.\d+",
with no minor version.

demerphq commented:
@leonerd you said someone objected to this, but nobody has comment on this PR, nor on the thread on p5p.  Can you ask them to comment here please?

demerphq commented:
> We discussed offering split-up deprecation categories, so you can "no warnings 'deprecated::.xyz" and re-affirmed that we want to do this.

Based on the PSC "PSC #101, 2023-03-17" comment above, and the fact noone has raised any objections, I am merging this.
